tlatools/org.lamport.tlatools/src/tlc2/TraceExplorer.java

```diff
@@ -147,31 +178,82 @@ public static void writeSpecTEStreams(
 		 * and CONSTANTS verbatim from the existing config file with which TLC ran. The
 		 * definition that appear in the .tla file (e.g. MC.tla) don't have to be copied
 		 * because SpecTE extends MC.
+		 * We also add the model values as constants so they can be used for the traces at
+		 * the TLA modules.
 		 * (see TraceExplorerDelegate#writeModelInfo)
 		 */
-		writer.addConstants(constants);
+		List<String> mvsConfigConstants = new ArrayList<String>();
+		for (String mv : mvsStr) {
+			mvsConfigConstants.add(SpecTraceExpressionWriter.indentString(String.format("%s = %s", mv, mv), 1));
+		}			
+
+		if(!constants.isEmpty()) {
+			List<String> indentedConstants = new ArrayList<String>();
+			// Add `CONSTANTS` header.
+			indentedConstants.add(TLAConstants.KeyWords.CONSTANTS);			
+			for (Vect<Object> keyValuePair : (ArrayList<Vect<Object>>) Collections.list(constants.elements())) {
+				String key = keyValuePair.elementAt(0).toString();
+				String value = keyValuePair.elementAt(1).toString();
+				indentedConstants.add(SpecTraceExpressionWriter.indentString(String.format("%s = %s", key, value), 1));
+			}
+
+			indentedConstants.addAll(mvsConfigConstants);
+			writer.addConstants(indentedConstants);
+		}
+
+		// If needed, create module which contain the reified constants.
+		// First we need to handle the case where a model value is defined in
+		// the config file, but, for some reason, it's not used in any of the
+		// model specs. E.g. I define `p1 = p1` in the `CONSTANTS` header of the
+		// config file, but I do not use it (through `CONSTANTS`) in the model spec.
+		// So we add these remaining constants to `mvsStr` so they are added to the 
+		// `TE` spec.
+		Defns defns = specInfo.getSpecProcessor().getDefns();
+		ArrayList<String> modConstants = new ArrayList<String>();
+		for (ModelValue mv : ModelValue.mvs) {
+			if (defns.get(mv.toString()) == null) {
+				modConstants.add(mv.toString());
+			}
+		}
+		final String teConstantSpecName = String.format("%s_%s", originalSpecName, TLAConstants.TraceExplore.SPEC_TECONSTANTS_NAME);
+		final Set<String> teConstantModuleHashSet = new HashSet<>();
+		String modelValuesAsConstants;
+		if (!modConstants.isEmpty()) {
+			teConstantModuleHashSet.add(teConstantSpecName);
+			modelValuesAsConstants = String.format("CONSTANTS %s\n", String.join(", ", modConstants));
```

Comment:
I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [FS: Format string should use %n rather than n (VA_FORMAT_STRING_USES_NEWLINE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#fs-format-string-should-use-n-rather-than-n-va-format-string-uses-newline).
This format string includes a newline character (\n). In format strings, it is generally  preferable to use %n, which will produce the platform-specific line separator.

Reply:
Ok, bot.

Reply:
Happy to help.